File: helper/Sources/HelperCore/Wire.swift
// SPDX-License-Identifier: Apache-2.0
import Foundation

/// The helper's version, which is SimMirror's: SimMirror refuses a helper of another version.
public enum HelperVersion {
    public static let current = "1.2.0"
}

/// What a frame on the socket is.
public enum FrameKind: UInt8, Sendable, CaseIterable {
    /// SimMirror asks for something.
    case request = 1
    /// The answer to a request.
    case reply = 2
    /// A request that did not work, and why.
    case failure = 3
    /// One piece of a stream a request started: an H.264 access unit.
    case chunk = 4
}

/// One message on the socket: a kind, the request it belongs to, a JSON object and any bytes that follow it.
///
/// On the wire, every number is big-endian:
///
///     u32 length     bytes after this field
///     u8  kind
///     u32 id         the request this frame asks or answers
///     u32 json size
///     json           a UTF-8 JSON object; empty means {}
///     blob           the rest: a JPEG, an access unit, or nothing
public struct Frame: Equatable, Sendable {
    public var kind: FrameKind
    public var id: UInt32
    public var json: Data
    public var blob: Data

    public init(kind: FrameKind, id: UInt32, json: Data = Data(), blob: Data = Data()) {
        self.kind = kind
        self.id = id
        self.json = json
        self.blob = blob
    }
}

/// Why bytes on the socket are not a frame.
public enum WireError: Error, Equatable, CustomStringConvertible {
    case tooLarge(Int)
    case unknownKind(UInt8)
    case jsonOverruns(Int)

    public var description: String {
        switch self {
        case .tooLarge(let size): return "a frame of \(size) bytes is larger than \(Wire.maxFrame)"
        case .unknownKind(let kind): return "frame kind \(kind) is not one this helper knows"
        case .jsonOverruns(let size): return "a frame's JSON of \(size) bytes runs past its end"
        }
    }
}

public enum Wire {
    /// The protocol's version, which SimMirror checks in the hello.
    public static let version = 1
    /// The bytes after the length a frame has before its JSON: kind, id and JSON size.
    static let headerAfterLength = 9
    /// The largest frame either side accepts.
    public static let maxFrame = 64 << 20

    public static func encode(_ frame: Frame) -> Data {
        var out = Data(capacity: 4 + headerAfterLength + frame.json.count + frame.blob.count)
        append(UInt32(headerAfterLength + frame.json.count + frame.blob.count), to: &out)
        out.append(frame.kind.rawValue)
        append(frame.id, to: &out)
        append(UInt32(frame.json.count), to: &out)
        out.append(frame.json)
        out.append(frame.blob)
        return out
    }

    static func append(_ value: UInt32, to data: inout Data) {
        withUnsafeBytes(of: value.bigEndian) { data.append(contentsOf: $0) }
    }

    static func uint32(_ data: Data, at offset: Int) -> UInt32 {
        let start = data.startIndex + offset
        return data[start..<start + 4].reduce(0) { ($0 << 8) | UInt32($1) }
    }
}

/// Frames out of a byte stream that arrives in whatever pieces the socket gives.
public struct FrameDecoder: Sendable {
    private var buffer = Data()

    public init() {}

    /// The frames `bytes` completes, in order. Throws when the stream is not frames, after which it cannot be read on.
    public mutating func push(_ bytes: Data) throws -> [Frame] {
        buffer.append(bytes)
        var frames: [Frame] = []
        while buffer.count >= 4 {
            let length = Int(Wire.uint32(buffer, at: 0))
            guard length >= Wire.headerAfterLength, length <= Wire.maxFrame else { throw WireError.tooLarge(length) }
            guard buffer.count >= 4 + length else { break }
            let rawKind = buffer[buffer.startIndex + 4]
            guard let kind = FrameKind(rawValue: rawKind) else { throw WireError.unknownKind(rawKind) }
            let id = Wire.uint32(buffer, at: 5)
            let jsonSize = Int(Wire.uint32(buffer, at: 9))
            guard jsonSize <= length - Wire.headerAfterLength else { throw WireError.jsonOverruns(jsonSize) }
            let jsonStart = buffer.startIndex + 13
            let end = buffer.startIndex + 4 + length
            frames.append(
                Frame(
                    kind: kind,
                    id: id,
                    json: Data(buffer[jsonStart..<jsonStart + jsonSize]),
                    blob: Data(buffer[jsonStart + jsonSize..<end])
                )
            )
            buffer = Data(buffer[end...])
        }
        return frames
    }
}

File: src/sim_mirror/connectors/native/wire.py
# SPDX-License-Identifier: Apache-2.0
"""The native helper's wire protocol, from SimMirror's side.

Every message on the helper's socket is a frame, and every number in it is big-endian::

    u32 length     bytes after this field
    u8  kind       REQUEST, REPLY, FAILURE or CHUNK
    u32 id         the request this frame asks or answers
    u32 json size
    json           a UTF-8 JSON object; empty means {}
    blob           the rest: a JPEG, an access unit, or nothing

SimMirror sends requests (``{"op": "screenshot", ...}``); the helper answers each with a reply, a failure that says
why, or -- for a stream -- chunks until the connection closes. The helper's side is
``helper/Sources/HelperCore/Wire.swift``, and both are tested against the same vectors
(``helper/Tests/Fixtures/wire-vectors.json``).
"""

from __future__ import annotations

import asyncio
import json
import struct
from dataclasses import dataclass, field
from typing import Any

#: The protocol's version, which a hello must say.
VERSION = 1

REQUEST = 1
REPLY = 2
FAILURE = 3
CHUNK = 4
KINDS = frozenset({REQUEST, REPLY, FAILURE, CHUNK})

#: The largest frame either side accepts.
MAX_FRAME = 64 << 20
_HEADER = struct.Struct(">IBII")
#: The bytes a frame's length counts before its JSON: kind, id and JSON size.
_AFTER_LENGTH = _HEADER.size - 4


class WireError(Exception):
    """Bytes on the helper's socket that are not a frame."""


@dataclass(frozen=True)
class Frame:
    kind: int
    id: int
    json: bytes = b""
    blob: bytes = field(default=b"", repr=False)

    def document(self) -> dict[str, Any]:
        """The JSON object the frame carries; an empty one is {}."""
        if not self.json:
            return {}
        try:
            value: object = json.loads(self.json)
        except ValueError as exc:
            raise WireError(f"a frame whose JSON cannot be read: {exc}") from exc
        if not isinstance(value, dict):
            raise WireError("a frame whose JSON is not an object")
        return value


def encode(frame: Frame) -> bytes:
    return _HEADER.pack(_AFTER_LENGTH + len(frame.json) + len(frame.blob), frame.kind, frame.id, len(frame.json)) + (
        frame.json + frame.blob
    )


def request(request_id: int, op: str, **fields: Any) -> bytes:
    """A request frame asking for `op` with these fields."""
    body = json.dumps({"op": op, **fields}, separators=(",", ":"), sort_keys=True).encode()
    return encode(Frame(REQUEST, request_id, body))


async def read_frame(reader: asyncio.StreamReader) -> Frame | None:
    """The next frame, or None when the helper has closed the connection between frames."""
    try:
        header = await reader.readexactly(_HEADER.size)
    except asyncio.IncompleteReadError as exc:
        if not exc.partial:
            return None
        raise WireError("the helper closed the connection inside a frame") from exc
    length, kind, frame_id, json_size = _HEADER.unpack(header)
    if length < _AFTER_LENGTH or length > MAX_FRAME:
        raise WireError(f"a frame of {length} bytes")
    if kind not in KINDS:
        raise WireError(f"frame kind {kind} is not one SimMirror knows")
    if json_size > length - _AFTER_LENGTH:
        raise WireError(f"a frame's JSON of {json_size} bytes runs past its end")
    try:
        body = await reader.readexactly(length - _AFTER_LENGTH)
    except asyncio.IncompleteReadError as exc:
        raise WireError("the helper closed the connection inside a frame") from exc
    return Frame(kind, frame_id, body[:json_size], body[json_size:])
